grafonnext/mkschema.go

```diff
@@ -0,0 +1,73 @@
+package main
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"cuelang.org/go/cue/ast"
+	"github.com/grafana/grafana/pkg/cuectx"
+	"github.com/grafana/grafana/pkg/plugins/pfs/corelist"
+	"github.com/grafana/grafana/pkg/registry/corekind"
+	"github.com/grafana/thema/encoding/jsonschema"
+)
+
+// this is a meaningless hack, we have to work out versioning
+var grafanaVersion = "latest"
+
+var rt = cuectx.GrafanaThemaRuntime()
+
+func main() {
+	// All core kinds first
+	base := filepath.Join("schemas", grafanaVersion, "core")
+	die(os.MkdirAll(base, 0777))
+
+	for _, kind := range corekind.NewBase(nil).AllStructured() {
+		f, err := jsonschema.GenerateSchema(kind.Lineage().Latest())
+		if err != nil {
+			die(fmt.Errorf("%s: %w", kind.Name(), err))
+		}
+
+		b := marshalindent(f)
+		path := filepath.Join(base, kind.MachineName()+".json")
+		die(os.WriteFile(path, b, 0644))
+	}
+
+	// Now, all composable kinds from plugins. (Note that core grafana hasn't been refactored
+	// to use 'composable kind' naming for its various methods yet)
+	for _, ptree := range corelist.New(nil) {
+		for slot, lin := range ptree.RootPlugin().SlotImplementations() {
+			f, err := jsonschema.GenerateSchema(lin.Latest())
+			if err != nil {
+				// There's a bug in generating openapi from Thema with imports are loaded.
+				// Normally an edge case in Thema, but needed in Grafana. Commenting this out
+				// until bug is fixed and just dropping errors instead
+				// die(fmt.Errorf("%s: %w", ptree.RootPlugin().Meta().Id, err))
```

Comment:
If this is uncommented, then we get errors like the following:

```
barchart: package "github.com/grafana/grafana/packages/grafana-schema/src/schema" imported but not defined in  (and 5 more errors)
```

This happens because of a [fugly hack](https://github.com/grafana/thema/blob/main/internal/util/util.go#L70) i have to do in Thema in order to be able to generate OpenAPI at all, because [the OpenAPI encoder uses the old, deprecated `cue.Instance`](https://pkg.go.dev/cuelang.org/go/encoding/openapi#Generate). Basically, that func writes out a text version of the lineage and recompiles it through a backdoor - but when recompiling, external imports can't be sanely satisfied.

They fixed the API in the recent alpha release, so i no longer have to export, but...well, there's another problem, and i probably just need to submit a patch upstream. Hoping to do that this week - this is also a prerequisite for grok, so it needs to get done regardless.